ocaml/xapi-guard/test/cache_test.ml

```diff
@@ -0,0 +1,139 @@
+let ( let@ ) f x = f x
+
+let ( let* ) = Lwt.bind
+
+module Tpm = Xapi_guard.Types.Tpm
+
+module TPMs = struct
+  let request_persist uuid write =
+    let __FUN = __FUNCTION__ in
+
+    let key = Tpm.deserialize_key (Random.int 3) in
+
+    let time = Mtime_clock.now () in
+    let contents =
+      Printf.sprintf "contents %s" (Mtime.to_uint64_ns time |> Int64.to_string)
+    in
+    let* () =
+      Logs_lwt.app (fun m ->
+          m "%s: New request! %a/%i/%a" __FUN Uuidm.pp uuid
+            Tpm.(serialize_key key)
+            Mtime.pp time
+      )
+    in
+    write (uuid, time, key) contents
+end
+
+let lwt_reporter () =
+  let buf_fmt ~like =
+    let b = Buffer.create 512 in
+    ( Fmt.with_buffer ~like b
+    , fun () ->
+        let m = Buffer.contents b in
+        Buffer.reset b ; m
+    )
+  in
+  let app, app_flush = buf_fmt ~like:Fmt.stdout in
+  let dst, dst_flush = buf_fmt ~like:Fmt.stderr in
+  let reporter = Logs_fmt.reporter ~app ~dst () in
+  let report src level ~over k msgf =
+    let k () =
+      let write () =
+        match level with
+        | Logs.App ->
+            Lwt_io.write Lwt_io.stdout (app_flush ())
+        | _ ->
+            Lwt_io.write Lwt_io.stderr (dst_flush ())
+      in
+      let unblock () = over () ; Lwt.return_unit in
+      Lwt.finalize write unblock |> Lwt.ignore_result ;
+      k ()
+    in
+    reporter.Logs.report src level ~over:(fun () -> ()) k msgf
+  in
+  {Logs.report}
+
+let setup_log level =
+  Logs.set_level level ;
+  Logs.set_reporter (lwt_reporter ()) ;
+  ()
+
+let ok = Lwt_result.ok
+
+let retry_forever fname f =
+  let rec loop () =
+    let* () =
+      Lwt.catch f (function exn ->
+          let* () =
+            Logs_lwt.app (fun m ->
+                m "%s failed with %s, retrying..." fname (Printexc.to_string exn)
```

Comment:
The stacktrace would be lost here (although in the case of Lwt it is mostly useless). Still might be useful to log it, just in case it contains something useful. To do that capture the raw stacktrace as soon as you enter the catch.